lib/streamlit/runtime/fragment.py

```diff
@@ -105,9 +108,73 @@ def fragment(
 
 
 @gather_metrics("experimental_fragment")
-def fragment(  # type: ignore[empty-body]
+def fragment(
     func: F | None = None,
     *,
     run_every: float | None = None,
 ) -> Callable[[F], F] | F:
-    pass
+    """TODO(vdonato): Write a docstring for this function."""
+
+    if func is None:
+        # Support passing the params via function decorator
+        def wrapper(f: F) -> F:
+            return fragment(
+                func=f,
+                run_every=run_every,
+            )
+
+        return wrapper
+    else:
+        non_optional_func = func
+
+    @wraps(non_optional_func)
+    def wrap(*args, **kwargs):
+        import streamlit as st
+
+        ctx = get_script_run_ctx()
+        if ctx is None:
+            return
+
+        if len(dg_stack.get()) > 0:
```

Comment:
nit: If I understand that correctly here, it checks if it is already nested inside any kind of container, we just use this container. Otherwise, we wrap it into a new container. or? Maybe add a comment here to explain that. 
Also, why don't we always just put it into a dedicated container?

Reply:
Hm let me just add a TODO comment here for now to investigate whether we can just always add a dedicated container. I vaguely remember running into weird difficulties doing this while working on the prototype that I didn't diagnose at the time

Reply:
If we don't wrap it, we probably need to do some more extensive testing with scenarios where elements are rendered above and underneath the partial, e.g.:

```
with st.expander("expander"): 
   # Some elements above
    my_fragment()
    # Some elements underneath
```